thrimshim/thrimshim/main.py

```diff
@@ -607,6 +608,171 @@ def _write_audit_log(conn, ident, api_action, editor, old_row=None, new_row=None
+@request_stats
+def list_templates():
+	"""List names of thumbnail templates in the database."""
+	with app.db_manager.get_conn() as conn:
+		query = """
+			SELECT name, description, attribution, crop, location FROM templates ORDER BY name
+		"""
+		results = database.query(conn, query)
+		return json.dumps([row._asdict() for row in results.fetchall()])
+
+
+@app.route('/thrimshim/template/<name>.png')
+@request_stats
+def get_template(name):
+	"""Get a thumbnail template in PNG form"""
+	with app.db_manager.get_conn() as conn:
+		query = """
+			SELECT image FROM templates WHERE name = %s 
+		"""
+		results = database.query(conn, query, name)
+		row = results.fetchone()
+		if row is None:
+			return 'Template {} not found'.format(name), 404
+
+		image = row[0]
+		return flask.Response(image, mimetype='image/png')
+
+
+@app.route('/thrimshim/template-metadata/<name>')
+@request_stats
+def get_template_metadata(name):
+	"""Get the metadata for a thumbnail as JSON"""
+	with app.db_manager.get_conn() as conn:
+		query = """
+			SELECT name, description, attribution, crop, location FROM templates WHERE name = %s 
+		"""
+		results = database.query(conn, query, name)
+		row = results.fetchone()
+		if row is None:
+			return 'Template {} not found'.format(name), 404
+		return json.dumps(row._asdict())
+
+
+@app.route('/thrimshim/add-template', methods=['POST'])
+@request_stats
+def add_template(artist=None):
+	"""Add a template to the database"""
+	new_template = flask.request.json
+
+	columns = ['name', 'image', 'description', 'attribution', 'crop', 'location']
+	#check for missing fields
+	missing = set(columns) - set(new_template)
+	if missing:
+		return 'Fields missing in JSON: {}'.format(', '.join(missing)), 400
+	# delete any extras
+	extras = set(new_template) - set(columns)
+	for extra in extras:
+		del new_template[extra]
+
+	#convert and validate template image
+	try:
+		new_template['image'] = base64.b64decode(new_template['image'])
+	except binascii.Error:
+		return 'Template image must be valid base64', 400
+	# check for PNG file header
+	if not new_template['thumbnail_image'].startswith(b'\x89PNG\r\n\x1a\n'):
+		return 'Template image must be a PNG', 400	
+
+	with app.db_manager.get_conn() as conn:
+
+		#check if name is already in the database
+		query = sql.SQL("""
+			SELECT name FROM events WHERE name = %s 
+		""")
+		results = database.query(conn, query, new_template['name'])
+		if results.fetchone() is not None:
+			return 'Template with name {} already exists'.format(new_template['name']), 400
+
+		query = sql.SQL("""
+			INSERT INTO templates ({})
+			VALUES ({})
+		""").format(
+				sql.SQL(", ").join(sql.Identifier(column) for column in columns),
+				sql.SQL(", ").join(database.get_column_placeholder(column) for column in columns),
+			)
+		database.query(conn, query, **new_template)
+	
+	return ''
+
+@app.route('/thrimshim/update-template/<name>', methods=['POST'])
+@request_stats
+def update_template(name, artist=None):
```

Comment:
This function is 90% the same as the create function, could we split out and reuse the common parts for input parsing, etc?

Reply:
I am splitting the common input parsing off into its own function but keeping the database interactions in the individual functions.